File: Precise/src/main/java/de/gdietz/fun/fractal/util/BigComplex.java
package de.gdietz.fun.fractal.util;

import java.math.BigDecimal;
import java.math.MathContext;
import java.math.RoundingMode;

public class BigComplex extends BigCoordinate implements NormedNumber<BigComplex>, BigNormed<BigComplex> {

    public static final BigComplex ZERO = new BigComplex();
    public static final BigComplex ONE = new BigComplex(BigDecimal.ONE);
    public static final BigComplex I = new BigComplex(BigDecimal.ZERO, BigDecimal.ONE);
    public static final BigComplex MINUS_ONE = new BigComplex(-1.0);

    private static final BigDecimal THREE = new BigDecimal(3);

    public BigComplex() {
	}

    public BigComplex(BigDecimal real) {
        super(real, BigDecimal.ZERO);
    }

    public BigComplex(BigDecimal real, BigDecimal imag) {
        super(real, imag);
    }

    public BigComplex(BigCoordinate c) {
        super(c);
    }

	public BigComplex(double real) {
		this(new BigDecimal(real), BigDecimal.ZERO);
	}

	public BigComplex(double real, double imag) {
        super(real, imag);
	}

	public BigComplex(Coordinate c) {
		super(c);
	}

	public BigDecimal getReal() {return x; }
	public BigDecimal getImag() {return y; }

    public BigComplex negate() {
		return new BigComplex(x.negate(), y.negate());
	}

    public BigComplex add(BigDecimal r) {
        return new BigComplex(x.add(r), y);
    }

    public BigComplex add(double r) {
        return add(new BigDecimal(r));
    }

    public BigComplex add(BigComplex c) {
		return new BigComplex(x.add(c.x), y.add(c.y));
	}

    public BigComplex subtract(BigDecimal r) {
        return new BigComplex(x.subtract(r), y);
    }

    public BigComplex subtract(double r) {
        return subtract(new BigDecimal(r));
    }

    public BigComplex subtract(BigComplex c) {
		return new BigComplex(x.subtract(c.x), y.subtract(c.y));
	}

	public BigComplex multiply(BigDecimal r) {
		return new BigComplex(r.multiply(x), r.multiply(y));
	}

    public BigComplex multiply(double r) {
        return multiply(new BigDecimal(r));
    }

	public BigComplex multiply(BigComplex c) {
		return new BigComplex(x.multiply(c.x).subtract(y.multiply(c.y)), x.multiply(c.y).add(y.multiply(c.x)));
	}

	public BigComplex sqr() {
        BigDecimal h = x.multiply(y);
		return new BigComplex(x.multiply(x).subtract(y.multiply(y)), h.add(h));
	}

	public BigComplex cube() {
		BigDecimal x2 = x.multiply(x);
		BigDecimal y2 = y.multiply(y);
		return new BigComplex(x.multiply(x2.subtract(THREE.multiply(y2))), THREE.multiply(x2).subtract(y2).multiply(y));
	}

	public BigComplex inverse() {
		BigDecimal den = normSqrPrecise();
        BigDecimal factor = BigDecimal.ONE.divide(den);
		return conjugate().multiply(factor);
	}

    public BigComplex inverse(MathContext mc) {
        BigDecimal den = normSqrPrecise();
        BigDecimal factor = BigDecimal.ONE.divide(den, mc);
        return conjugate().multiply(factor);
    }

    public BigComplex divide(BigDecimal r) {
        BigDecimal factor = BigDecimal.ONE.divide(r);
        return multiply(factor);
    }

    public BigComplex divide(BigComplex c) {
        return multiply(c.inverse());
	}

    public BigComplex divide(BigDecimal r, MathContext mc) {
        BigDecimal factor = BigDecimal.ONE.divide(r, mc);
        return multiply(factor);
    }

    public BigComplex divide(BigComplex c, MathContext mc) {
        return multiply(c.inverse(mc));
	}

	public BigComplex conjugate() {
	    return new BigComplex(x, y.negate());
	}

	public BigComplex pow(int n) {
		if (n==2) return sqr();
		if (n==3) return cube();
		if (n==-1) return inverse();
		if (n==1) return this;
		if (n==0) return BigComplex.ONE;
		if (n<0) return inverse().pow(-n);
		if (n % 3 == 0) return cube().pow(n / 3);
		if (n % 2 == 0) return sqr().pow(n / 2);
		return multiply(sqr().pow((n-1) / 2));
	}

    public BigComplex getZero() {
        return ZERO;
    }

    public BigComplex getUnit() {
        return ONE;
    }

    public boolean isUnit() {
        return x.compareTo(BigDecimal.ONE) == 0 && y.compareTo(BigDecimal.ZERO) == 0;
    }

    public int scale() {
        return Math.max(x.scale(), y.scale());
    }

    public BigComplex setScale(int newScale, RoundingMode roundingMode) {
        return new BigComplex(x.setScale(newScale, roundingMode), y.setScale(newScale, roundingMode));
    }

    public BigComplex setScale(int newScale) {
        return new BigComplex(x.setScale(newScale), y.setScale(newScale));
    }

    public Complex complexValue() {
        return new Complex(x.doubleValue(), y.doubleValue());
    }

    public String toString(boolean parentheses, boolean negativeParentheses, boolean sign) {
        BigComplexNumberFormatter cnf = new BigComplexNumberFormatter(parentheses, negativeParentheses, sign);
        cnf.addNumberString(x);
        cnf.addNumberString(y, "i");
        return cnf.toString();
    }

    public String toString(boolean parentheses) {
        return toString(parentheses, false, false);
    }

    public String toString() {
        return toString(false);
    }

}

File: FractalScala/src/main/scala/de/gdietz/fun/fractal/scala/util/BigComplex.scala
package de.gdietz.fun.fractal.scala.util

import de.gdietz.fun.fractal.util.{BigComplexNumberFormatter, BigCoordinate, Coordinate, BigComplex => JavaBigComplex}

import java.math.{BigDecimal => JavaBigDecimal}
import scala.language.implicitConversions

sealed trait OptBigComplex
  extends OptHigherNumber[OptBigComplex, BigComplex]
    with Product with Serializable {

  override def none: OptBigComplex = NoBigComplex
  override def zero: BigComplex = BigComplex.zero
  override def unit: BigComplex = BigComplex.one

}

object OptBigComplex {

  def apply(x: BigDecimal, y: BigDecimal = BigReal.decimalZero): OptBigComplex =
    BigComplex(x, y)

  val none: OptComplex = NoComplex


  implicit def optComplexToOptBigComplex(c: OptComplex): OptBigComplex =
    c match {
      case Complex(x, y) => BigComplex(x, y)
      case _ => NoBigComplex
    }

  implicit def bigDecimalToOptBigComplex(x: BigDecimal): OptBigComplex =
    BigComplex(x)

  implicit def javaBigDecimalToOptBigComplex(x: JavaBigDecimal): OptBigComplex =
    BigComplex(x)

  implicit def doubleToOptBigComplex(x: Double): OptBigComplex =
    BigComplex(x)

  implicit def optRealToOptBigComplex(x: OptReal): OptBigComplex =
    x match {
      case Real(x) => BigComplex(x)
      case _ => NoBigComplex
    }

  implicit def optBigRealOptToBigComplex(x: OptBigReal): OptBigComplex =
    x match {
      case BigReal(x) => BigComplex(x)
      case _ => NoBigComplex
    }

  implicit def bigCoordinateToOptBigComplex(c: BigCoordinate): OptBigComplex =
    BigComplex(c.getX, c.getY)

  implicit def coordinateToOptBigComplex(c: Coordinate): OptBigComplex =
    BigComplex(c.getX, c.getY)

}


case class BigComplex(x: BigDecimal, y: BigDecimal = BigReal.decimalZero)
  extends OptBigComplex
    with SomeHigherNumber[OptBigComplex, BigComplex] {

  override def isZero: Boolean = x == BigReal.decimalZero && y == BigReal.decimalZero

  override def isUnit: Boolean = x == BigReal.decimalOne && y == BigReal.decimalZero

  def normSqrPrecise: BigDecimal = x * x + y * y

  @inline override def normSqr: Double = normSqrPrecise.toDouble

  override def unary_- : BigComplex = BigComplex(-x, -y)

  override def +(r: Double): BigComplex = {
    val r0 = BigDecimal.decimal(r)
    BigComplex(x + r0, y)
  }

  override def +(c: BigComplex): BigComplex = BigComplex(x + c.x, y + c.y)

  override def -(r: Double): BigComplex = {
    val r0 = BigDecimal.decimal(r)
    BigComplex(x - r0, y)
  }

  override def -:(r: Double): BigComplex = {
    val r0 = BigDecimal.decimal(r)
    BigComplex(r0 - x, -y)
  }

  override def -(c: BigComplex): BigComplex = BigComplex(x - c.x, y - c.y)

  override def *(r: Double): BigComplex = {
    val r0 = BigDecimal.decimal(r)
    BigComplex(r0 * x, r0 * y)
  }

  override def *(c: BigComplex): BigComplex = BigComplex(x * c.x - y * c.y, x * c.y + y * c.x)

  override def inverse: BigComplex = {
    val den = normSqrPrecise
    BigComplex(x / den, -y / den)
  }

  override def /(r: Double): BigComplex = {
    val r0 = BigDecimal.decimal(r)
    BigComplex(x / r0, y / r0)
  }

  override def /(c: BigComplex): BigComplex = {
    val den = c.normSqrPrecise
    BigComplex((x * c.x + y * c.y) / den, (y * c.x - x * c.y) / den)
  }

  override def conjugate: BigComplex = BigComplex(x, -y)

  override def sqr: BigComplex = {
    val h = x * y
    BigComplex(x * x - y * y, h + h)
  }

  override def cube: BigComplex = {
    val x2 = x * x
    val y2 = y * y
    BigComplex(x * (x2 - BigReal.decimalThree * y2), (BigReal.decimalThree * x2 - y2) * y)
  }

  @inline override def pow(n: Int): BigComplex = super.pow(n)

  @inline override def **(n: Int): BigComplex = pow(n)


  def toJavaBigComplex: JavaBigComplex = new JavaBigComplex(x.bigDecimal, y.bigDecimal)


  def toString(parentheses: Boolean = false,
               negativeParentheses: Boolean = false,
               sign: Boolean = false): String = {
    val cnf = new BigComplexNumberFormatter(parentheses, negativeParentheses, sign)
    cnf.addNumberString(x.bigDecimal)
    cnf.addNumberString(y.bigDecimal, "i")
    cnf.toString
  }

  override def toString: String = toString(parentheses = false)

}

object BigComplex {

  val zero: BigComplex = BigComplex(BigReal.decimalZero)
  val one: BigComplex = BigComplex(BigReal.decimalOne)
  val i: BigComplex = BigComplex(BigReal.decimalZero, BigReal.decimalOne)
  val minusOne: BigComplex = BigComplex(BigReal.decimalMinusOne)
  val minusI: BigComplex = BigComplex(BigReal.decimalZero, BigReal.decimalMinusOne)


  implicit def complexToBigComplex(c: Complex): BigComplex =
    BigComplex(c.x, c.y)

  implicit def bigDecimalToBigComplex(x: BigDecimal): BigComplex =
    BigComplex(x)

  implicit def javaBigDecimalToBigComplex(x: JavaBigDecimal): BigComplex =
    BigComplex(x)

  implicit def doubleToBigComplex(x: Double): BigComplex =
    BigComplex(x)

  implicit def realToBigComplex(x: Real): BigComplex =
    BigComplex(x.x)

  implicit def bigRealToBigComplex(x: BigReal): BigComplex =
    BigComplex(x.x)

  implicit def bigCoordinateToBigComplex(c: BigCoordinate): BigComplex =
    BigComplex(c.getX, c.getY)

  implicit def coordinateToBigComplex(c: Coordinate): BigComplex =
    BigComplex(c.getX, c.getY)

}

case object NoBigComplex
  extends OptBigComplex
    with NoHigherNumber[OptBigComplex, BigComplex] {

  override def toString: String = "na"

}
